components/config/src/config/mod.rs

```diff
@@ -206,18 +206,30 @@ impl Config {
 
     /// Merges the extra data from the theme with the config extra data
     fn add_theme_extra(&mut self, theme: &Theme) -> Result<()> {
-        // 3 pass merging
-        // 1. save config to preserve user
-        let original = self.extra.clone();
-        // 2. inject theme extra values
         for (key, val) in &theme.extra {
-            self.extra.entry(key.to_string()).or_insert_with(|| val.clone());
-        }
-
-        // 3. overwrite with original config
-        for (key, val) in &original {
-            self.extra.entry(key.to_string()).or_insert_with(|| val.clone());
-        }
+            if !self.extra.contains_key(key) {
+                // The key is not overriden in site config, insert it
+                self.extra.insert(key.to_string(), val.clone());
+                continue;
+            }
+            if !val.is_table() {
+				// The key is overriden in site config, and is not a mapping (table)
+				// so we discard it (don't merge)
+                continue;
+            }
+			// The key is a table, so we want to merge its child entries with config
+			// Safe unwraps because we have checks before
+			let theme_table = val.as_table().unwrap();
+			let config_table = self.extra.get_mut(key).unwrap()
+				.as_table_mut().expect(&format!("config.toml and theme.toml have incompatible types for entry theme.extra.{}", key));
+
+			for (child_key, child_val) in theme_table {
+				if !config_table.contains_key(child_key) {
+					// Theme setting is not overriden, insert it
+					config_table.insert(child_key.to_string(), child_val.clone());
+				}
+			}
+		}
```

Comment:
It would need to be recursive to be fully working no?

Reply:
Yes, without recursion we only do one more level deep merging, so a subsubtable `[config.extra.foo.bar]` would not be merged properly. Should i make it correct and merge recursively?

Reply:
If it's not too much work, sure!